src/main/java/org/candlepin/subscriptions/subscription/SubscriptionTableController.java

```diff
@@ -0,0 +1,246 @@
+import org.candlepin.subscriptions.db.SubscriptionCapacityViewRepository;
+import org.candlepin.subscriptions.db.model.ServiceLevel;
+import org.candlepin.subscriptions.db.model.SubscriptionCapacityView;
+import org.candlepin.subscriptions.db.model.Usage;
+import org.candlepin.subscriptions.resource.ResourceUtils;
+import org.candlepin.subscriptions.util.ApplicationClock;
+import org.candlepin.subscriptions.utilization.api.model.*;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Component;
+
+@Component
+public class SubscriptionTableController {
+
+  private final SubscriptionCapacityViewRepository subscriptionCapacityViewRepository;
+  private final ApplicationClock clock;
+
+  SubscriptionTableController(
+      SubscriptionCapacityViewRepository subscriptionCapacityViewRepository,
+      ApplicationClock clock) {
+    this.subscriptionCapacityViewRepository = subscriptionCapacityViewRepository;
+    this.clock = clock;
+  }
+
+  public SkuCapacityReport capacityReportBySku( // NOSONAR
+      ProductId productId,
+      @Min(0) Integer offset,
+      @Min(1) Integer limit,
+      ServiceLevelType sla,
+      UsageType usage,
+      Uom uom,
+      SkuCapacityReportSort sort,
+      SortDirection dir) {
+    /*
+    Notes:
+    - Ignoring beginning and ending query params, as the table should (currently, anyway) only
+      report on the latest subscription status information.
+    - The implementation will report "subscription end" events for Next Event ONLY. That is,
+      future-dated subs are not taken into account when reporting the Next Event; only currently
+      active subs are used for calculating Next Event.
+    */
+
+    OffsetDateTime reportEnd = clock.now();
+    OffsetDateTime reportStart = reportEnd.minusYears(1);
+    ServiceLevel sanitizedServiceLevel = sanitizeServiceLevel(sla);
+    Usage sanitizedUsage = sanitizeUsage(usage);
+
+    List<SubscriptionCapacityView> capacities =
+        subscriptionCapacityViewRepository.findAllBy(
+            getOwnerId(),
+            productId.toString(),
+            sanitizedServiceLevel,
+            sanitizedUsage,
+            reportStart,
+            reportEnd);
+
+    Map<String, SkuCapacity> inventories = new HashMap<>();
+    for (SubscriptionCapacityView subscriptionCapacityView : capacities) {
+      String sku = subscriptionCapacityView.getSku();
+      final SkuCapacity inventory =
+          inventories.computeIfAbsent(
+              sku, key -> initializeDefaultSkuCapacity(subscriptionCapacityView, uom));
+      calculateNextEvent(subscriptionCapacityView, inventory, reportEnd);
+      addSubscriptionInformation(subscriptionCapacityView, inventory);
+      addTotalCapacity(subscriptionCapacityView, inventory);
+    }
+
+    List<SkuCapacity> reportItems = new ArrayList<>(inventories.values());
+    Pageable pageable = ResourceUtils.getPageable(offset, limit);
+    reportItems = paginate(reportItems, pageable);
```

Comment:
Can you please leave a note that this is an area of potential performance improvement in the future (by paging at the DB)?

Reply:
Added a comment.